### DISCUSSION TITLE
ncurses buildable with dj64dev

### DISCUSSION THREAD

**Discussion by stsp** [dosemu2/dosemu2#2289] [General]
After a bit of work I have this:
https://github.com/stsp/ncurses
All the needed patches (just 2
actually) are already mailed to
ncurses author as they are
needed for djgpp as well.

Had to rework dj64dev slightly so
that configure script can deal with
it. But at the end, all you need for
building something with dj64dev,
is just this:
https://github.com/stsp/ncurses/blob/master/configure-dj64
I never though hooking it to autoconf
can be made that simple. :)
After building, you may find that
all tests segfault. This is intended:
such a simple configure wrapper
can't hook in the linking process.
So I added another 1-line script:
https://github.com/stsp/ncurses/blob/master/test/link2exe.sh
You can do:
`./link2exe.sh dots` and it will link
`dots.exe` which can be run this way:
`dosemu ./dots.exe TERM=ansi TERMINFO_DIRS=c:/terminfo`
You need `c:\terminfo\a\ansi` for this
to work. Terminfo files can be copied
from `/usr/share/terminfo`. `ansi` is
not working well, gives lots of artefacts.
Perhaps some other terminfo file can
do better. So far I don't know how this
all works, i.e. this `ansi` terminal doesn't
even need `nansi.sys`, which is strange.
But the fact that it all works, and hooks
to autoconf that easily, is more than
amazing. Also I used host gdb to nail
down the path separator bug:
https://github.com/stsp/ncurses/commit/3d66a9ac733db1d2e92abb1dc9b9df3fd22de38d
Not sure how much it would take for
the one to debug this under djgpp
itself...

Maybe @jwt27 can also be interested
in that ncurses fork, as it works with
djgpp as well.

**Comment by stsp** [comment]
Digging a bit deeper, it appears that
djgpp itself has a built-in terminal
emulator. So the proper terminfo
entry is not `a/ansi`, but rather
`d/djgpp204`.

**Comment by stsp** [comment]
Instead of mangling the ncurses build
system, I implemented a kind of go64
extender that can load ELF files directly.
Also ncurses upstream seem to have applied
all the needed patches:
https://lists.gnu.org/archive/html/bug-ncurses/2024-10/msg00019.html
(I bargained them as a djgpp-required
patches, but they were needed for dj64dev)

I guess cross-compilation for DOS was
never that simple. And ELF is going to
be the next DOS file format, it seems. :)
With that you don't even need `djlink` -
everything is done by host tools.

**Reply by andrewbird** [reply]
So trying to get my head around this. What can an ELF format executable give you that an MZ can't? Would it be possible to run an unmodified ELF executable on a suitably modified DOS kernel on real hardware?

**Reply by stsp** [reply]
> What can an ELF format executable give you that an MZ can't?

Only that you can use dj64dev
with autoconf, w/o mangling the
build system. And that works only
if the program doesn't have any
x86 asm files. ncurses fortunately
has none. With asm files you can't
avoid djlink step.

> Would it be possible to run an unmodified ELF executable on a suitably modified DOS kernel on real hardware?

Not the ones produced with dj64dev.
Those produced with ia16-gcc - likely yes.

**Reply by stsp** [reply]
dj64dev produces 64bit `ET_DYN` elfs.
No hopes to ever run them under DOS,
except by embedding dosemu2 into an
extender (something like DSL, https://github.com/haileys/doslinux)

ia16-gcc AFAIK produces 32bit `ET_EXEC`
elfs. And they are statically linked. So I
don't see any problems supporting those
(they use some dodgy undocumented
relocations, but this is not a big problem)

**Reply by andrewbird** [reply]
> Not the ones produced with dj64dev.
Those produced with ia16-gcc - likely yes.

Is it possible to produce a dual mode ELF executable, i.e. one that could run on both dosemu and real hardware? But I guess even then the file size would become a barrier to uptake of the new format.

**Reply by stsp** [reply]
The problem is not in dual or
single mode. The problem is in
an incredible complexity of running
a dj64dev-produced elf. Even on
freebsd I had to limit the stuff to
statically-linked ELFs only. There are
no hopes to ever write such a loader
under DOS - its not impossible, but
nearly so. Double-format won't help:
what 64bit file format for DOS do you
know? I am not aware of any, so you
need to implement a loader for
64bit `ET_DYN` elfs. And that loader
itself needs to run on some 64bit kernel -
do you have one for DOS?
Its possible but its like writing an OS
from scratches.

OTOH you can support 32bit ET_EXEC
statically linked elfs (aka ia16-gcc) relatively
simply if you want.

**Reply by andrewbird** [reply]
I was trying to see how ELF binaries could become commonplace on DOS, so I guess it'll be Dosemu2 only that supports this format.

**Reply by stsp** [reply]
> I was trying to see how ELF binaries could become commonplace on DOS

You can use ELF for dos, but as
soon as you deal with 64bit elf,
you just need to write too many
software (64bit kernel+very complex
loader).

No problems with 32bit elfs, and
ET_EXEC+static linking makes it really
rather simple. But if you follow ia16-gcc,
then you need to employ their non-standard
relocs. This makes you incompatible with
eg 32bit elfs produced by nasm-segelf.
But being incompatible with nasm-segelf
is not a big problem, so you really CAN use
elf under DOS.

Unfortunately every tool-chain produces
its own ELF. So the problem is this: supporting
ELF doesn't give you any common ground
at all. You need to decide what tool-chain
you are going to support. And this is
unfortunate.

**Reply by stsp** [reply]
Or maybe you want to support DJELF -
djgpp elf support by Daniel Borca?
He did it, wrote an elf loader for DOS.
But once again, that loader only supports
his own ELFs.

**Reply by andrewbird** [reply]
I found this https://www.bttr-software.de/forum/mix_entry.php?id=202&page=0&order=subject&category=8 but the link inside is dead.

**Reply by stsp** [reply]
Yes, but it was up last year.
So I have his code downloaded
somewhere. Other people likely
have it too. If you are interested,
I can (try to) find his code for you.

Also, besides djelf, he wrote another
ELF loader called DXE. You can
find its sources in `libc/dxe` of
either dj64dev or djgpp repo.
I think it deals with 32bit ET_DYN elfs.

**Reply by andrewbird** [reply]
> If you are interested,
I can (try to) find his code for you.

Don't worry it's only a passing interest for me, but thanks anyway.

**Reply by stsp** [reply]
https://web.archive.org/web/20091026234423/http://geocities.com/dborca/djgpp/elf/src/

**Reply by andrewbird** [reply]
The wayback machine has it (thankfully those hacker B*ds didn't kill it) https://web.archive.org/web/20091027102102/http://www.geocities.com/dborca/djgpp/elf/djelf.html

**Reply by andrewbird** [reply]
**snap**!

**Reply by stsp** [reply]
So it was really a major attempt
to modernize djgpp. :) Not even
nearly at a scale of dj64dev, but
still a great work.

But as his description says, these
elfs are only capable of Flat mode
(aka djgpp). You can't use them
with segmented mode - for that
you need either nasm-segelf or
ia16-gcc ELFs.

**Reply by stsp** [reply]
After all that you may wonder why
.exe format is that capable? Why it
doesn't have all of the above problems?
And the fact is... it does! It has all of
them, and more.
But the solution is simple: it embeds
the realmode loader and the payload.
The payload is either COFF or LE or
whatever the loader understands.
Such solution is undesirable for ELF.
ELF is supposed to be the universal
format, so embedding the OS-specific
loader and some other format into it,
while possible, is considered a bad
practice. So the existing tool-chains
never do this for ELF, and always do
for .exe.

**Reply by stsp** [reply]
Here is another example.
Suppose you have some 32bit
executable elf produced by gcc
on a 32bit linux machine. What can
you do with it on a 64bit machine?
It will be unable to find the needed
loader. You can install multilib, which
will bring in the 32bit loader and all
the 32bit environment! The needed
loader will be found by looking into
the .interp section of an elf. So then
you can load it.

Now a more problematic case: suppose
you have a 32bit shared library elf on
a 64bit machine. What can you do with
it? You can't `dlopen()` it from your 64bit
program, no matter what you install.
So in this case you write your own elf
loader and a small hypervisor that can
embed the 32bit run-time into your
64bit process. I believe this is what
the recent versions of wine do. Or,
alternatively, you install multilib and
rebuild your prog for 32bit environment,
in which case you can simply `dlopen()`
your 32bit solib. This is what older
wine versions did.

So being elf, doesn't make your prog universal.
Even under linux, you have major troubles
running non-native elf executable, and absolutely
tremendous troubles loading non-native elf solib.
So even on linux, only the elfs created by the
particular (native) tool-chain are supported, and
if you use some `-m32` flag, you immediately
get a non-loadable elf, at least not w/o major troubles.
And dj64dev presents you with the worst possible
cases in all fronts when you want to load its elf file.
If you want to have elfs under DOS - support other
toolchains, not this one. :) Then you can.

The amount of efforts needed to run non-native
elf, can probably be estimated from freebsd's linuxulator.
Its a great effort, but even in this case the bitness
must match. When you have also bitness mismatch,
multiple that by 10. When you want to deal with
solibs of different bitness - multiple by 100. And
when you want to get that all under DOS - multiple
by a thousand. :)

**Reply by stsp** [reply]
I guess I can summarize that properly one day. :)
So in linux you have just 1 (or 2 in case of multilib)
elf loader: ld-linux.so.2. It only supports the native
GNU tool-chain, and nothing else is ever needed.
Once you have an elf file referring to a different
loader (eg built by another tool-chain), it doesn't load.

In DOS you can implement a basic elf support
in freedos kernel in a few hours. But then you
need to write the loaders. Many of them - one
per the tool-chain you want to support. Some
are easy, some are not. But the main problem
is that there is no single "native" elf tool-chain
for DOS, which you could support and say "done",
as is the case in linux.
You likely can easily write loaders for OW and
ia16-gcc, but not for dj64dev. And OW itself
will likely need to get tweaked to produce an
elf output (no idea if it supports any ELF targets
currently).

So to summarize: what makes elf _look_ universal,
is only the fact that linux gets away with having
just 1 loader for just 1 tool-chain. Other than that,
elf is just a container format where you can store
any trash and let the loader sort it out.

**Reply by stsp** [reply]
And in fact, I have convinced myself
that the current dj64dev's approach
can be tightened up:
https://github.com/stsp/dj64dev/issues/15
If I decided to use `ET_DYN` which is
already a huge portability problem,
then at least I can try `-nostdlib` so
that you don't have to port also entire
glibc under DOS. :)

**Reply by stsp** [reply]
Also filled
https://github.com/stsp/dj64dev/issues/16
https://github.com/stsp/dj64dev/issues/17
If they are implemented then we
have an optional support for ET_EXEC
elfs. It is almost feasible to use them
under DOS. Of course the fact that they
are 64bit, still leaves a lot of troubles.
But at least it can be moved from
needing a completely undocumented
custom loader, to any loader that
follows an ELF specs carefully enough.
Because when you deal with the statically
linked non-pie ET_EXEC elf, you don't
have to support every tool-chain separately:
such ELFs can be supported all at once.
Which is probably exactly what you want.

**Reply by stsp** [reply]
https://github.com/stsp/dj64dev/issues/15
is now implemented. So at least
you don't have to port glibc to DOS. :)
As for side-stepping ET_DYN - I think
its a wontfix stuff.

But your question was initially flawed:

> I was trying to see how ELF binaries could become commonplace on DOS,
> so I guess it'll be Dosemu2 only that supports this format.

No, the support is in comcom64!
`elfexec` command works, you can
use it to run elfs produced by dj64dev
right here, right now! Why was I
saying its impossible? Well, I was
confused: its already done. :)
Probably the fact that I failed to
port it to comcom32 (https://github.com/stsp/dj64dev/issues/16)
made me so confused. In theory
its portable to comcom32, just why
to bother.
But the support for 32bit ELF files
can be easily done in both comcom32
and comcom64. Its only the 64bit elf
is a bit problematic for comcom32.

So if you want the ELF support in DOS,
don't modify the DOS kernel. Modify
comcom! You can even use libelf there,
I guess its buildable even with plain djgpp.
And problem solved.
In fact, do you have .bat files support in
DOS? You do, but only in command.com.
And no one complains. :)

**Reply by stsp** [reply]
The very similar discussion happened here:
https://sourceware.org/bugzilla/show_bug.cgi?id=31795
Especially funny moments are like this:

```
> > Why are glibc implementation details dictating how ld.bfd operates? Where
> 
> This is how ELF works on Linux.

Well, ELF is not Linux specific, though.
```

Or:
```
> does it say non-0 load addresses for static-PIE/PIE are not allowed in the
> ELF specification? If it is allowed, why is ld.bfd not allowing me to do it

Static PIE isn't the part of ELF spec.
```

So basically hacking on ELFs seems
to be what everyone does. But if you
want to have static non-PIE ET_EXEC,
I still believe its feasible and well
standardized.
Also I didn't know static PIE also uses
ET_DYN, but it appears so after reading
that discussion. Now I don't see any
reason to side-step ET_DYN: everyone
seem to exploit it for their own needs.

**Reply by stsp** [reply]
> But if you
> want to have static non-PIE ET_EXEC,

But OTOH IIRC ia16-gcc uses static PIE,
which was my main complain to their
proposal, and the motivator behind
nasm-segelf. So no, limiting to static
non-PIE can't work, as ia16-gcc already
uses PIE. And even the nasm-segelf
scheme, while is achievable with binutils,
still uses some non-standard extensions
(and therefore doesn't work with lld).
Hmm, so I guess currently there is no
any possibility but to support every tool-chain
individually, no possibility to limit the support
to some "simple" scheme (like static non-pie
was before nasm-segelf), and so on. Everything
was crippled. :)

**Comment by stsp** [comment]
> What can an ELF format executable give you that an MZ can't? Would it be possible to run an unmodified
> ELF executable on a suitably modified DOS kernel on real hardware?

Ok, so its time for you to
tell me. :)
[test.tar.gz](https://github.com/user-attachments/files/29244960/test.tar.gz)

In the attached archive you
can find comcom32.exe and
hello.elf. You can run it this
way:
```
dosemu ./comcom32.exe
```

... or this way:
```
dosbox ./comcom32.exe
```

As far as I can tell, it works
both ways in a similar fashion -
I can't find any differences.
Now you can do:
```
elfexec2 hello.elf
```

... and you should get
`Hello World!` from an ELF file.
Unfortunately, on dosbox-staging
I am getting an error at that
point, and dosbox-x completely
crashes (it doesn't feel well even
before elfexec2).
But on classic dosbox it works
perfectly AFAICT.

Would be good if someone can
test that on other emulators or
under real DOS, as I don't know
if my code is buggy or all dosboxes,
except classic one, are buggy.

@mcericdos maybe you can check
that out on real DOS? :)
